### PR TITLE
util, bin: make big numbers human-readable

### PR DESCRIPTION
**Pull request by overvenus**
Support:
 - File size(based on byte):
  - **B**: Byte
  - **KB**: Kilobyte, 1KB = 1024B
  - **MB**: Megabyte, 1MB = 1024KB
  - **GB**: Gigabyte, 1GB = 1024MB
  - **TB**: Terabyte, 1TB = 1024GB
  - **PB**: Petabyte, 1PB = 1024TB
 - Time(based on ms):
  - **ms**: One thousandth of one second, 
  - **s**: One second, 1s = 1000ms
  - **m**: One minute, 1m = 60s
  - **h**: One hour, 1h = 60m

#772

### CODE REVIEW

File: etc/config-template.toml [hunk 1]
```diff
@@ -26,22 +26,22 @@ prefix = "tikv"
 # notify capacity, 40960 is suitable for about 7000 regions.
 notify-capacity = 40960
 # Region heartbeat tick interval (ms) for reporting to pd. 
-pd-heartbeat-tick-interval = 5000
+pd-heartbeat-tick-interval = "5000ms"
```
Comment (ngaut) [inline etc/config-template.toml]: Could you remove quote ? so we can simply write 5000ms.

Comment (siddontang) [inline etc/config-template.toml]: toml may not support it.

Comment (overvenus) [inline etc/config-template.toml]: toml does not support something like `5000ms` or `10MB`.

File: src/util/config.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+const TIME_MAGNITUDE: i64 = 1000;
+const TIME_MS: i64 = UNIT;
+const TIME_S: i64 = TIME_MS * TIME_MAGNITUDE;
+
+#[allow(match_same_arms)]
+pub fn get_integer_by_string(size: &str) -> Result<i64, ()> {
+    let (num, unit) = try!(split_property(size));
+
+    match unit {
```
Comment (siddontang) [inline src/util/config.rs]: lower the unit, so that we can use kb, mb too.

File: src/util/config.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        // file size
+        "B" => Ok((num * (FILE_SIZE_B as f64)) as i64),
```
Comment (siddontang) [inline src/util/config.rs]: I think B is needless

File: src/util/config.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                let res = num * (FILE_SIZE_GB as f64);
+                ret += res as i64;
+                decimal += res - ((res as i64) as f64);
+            }
+            Ok(ret + (decimal as i64))
+        }
+
+        // time
+        "ms" => Ok((num * (TIME_MS as f64)) as i64),
+        "s" => Ok((num * (TIME_S as f64)) as i64),
```
Comment (siddontang) [inline src/util/config.rs]: can we support m and h too?

File: src/util/config.rs (hunk 3 above)
Comment (overvenus) [inline src/util/config.rs]: I think we should keep `B`. People may get confused as they can write something like "10KB", "1.5MB", but why not "50B"?

Comment (siddontang) [inline src/util/config.rs]: Supporting 50B is strange, why not using 50 directly?
I see that https://raw.githubusercontent.com/antirez/redis/2.8/redis.conf doesn't support B too.

File: src/util/config.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    match num.parse::<f64>() {
+        Ok(f) => Ok((f, unit)),
+        Err(_) => Err(()),
+    }
+}
+
+const UNIT: i64 = 1;
+const FILE_SIZE_MAGNITUDE: i64 = 1024;
+const FILE_SIZE_B: i64 = UNIT;
+const FILE_SIZE_KB: i64 = FILE_SIZE_B * FILE_SIZE_MAGNITUDE;
```
Comment (siddontang) [inline src/util/config.rs]: I think following is more simple and clear. 
```
const KB = 1024
const MB = 1024 * 1024
```

File: src/util/config.rs (hunk 3 above)
Comment (overvenus) [inline src/util/config.rs]: OK, I will remove B

File: src/util/config.rs (hunk 5 above)
Comment (c4pt0r) [inline src/util/config.rs]: +1

File: src/util/config.rs (hunk 2 above)
Comment (c4pt0r) [inline src/util/config.rs]: +1

File: src/util/config.rs (hunk 3 above)
Comment (overvenus) [inline src/util/config.rs]: remove `ms` too? @siddontang

Comment (siddontang) [inline src/util/config.rs]: `ms`  is commonly used, I think ok.

**Comment by c4pt0r** [comment]
Well done，LGTM.

File: src/util/config.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    let (num, unit) = try!(split_property(size));
+
+    match &*unit.to_owned().to_lowercase() {
+        // file size
+        "kb" => Ok((num * (KB as f64)) as i64),
+        "mb" => Ok((num * (MB as f64)) as i64),
+        "gb" => Ok((num * (GB as f64)) as i64),
+        "tb" => {
+            let mut ret: i64 = 0;
+            let mut decimal: f64 = 0.0;
```
Comment (siddontang) [inline src/util/config.rs]: why using decimal here?

File: src/util/config.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+// PB = PB * DATA_MAGNITUDE
+
+const TIME_MAGNITUDE_1: i64 = 1000;
+const TIME_MAGNITUDE_2: i64 = 60;
+const MS: i64 = UNIT;
+const SECOND: i64 = MS * TIME_MAGNITUDE_1;
+const MINTUE: i64 = SECOND * TIME_MAGNITUDE_2;
+const HOUR: i64 = MINTUE * TIME_MAGNITUDE_2;
+
+pub fn get_integer_by_string(size: &str) -> Result<i64, ()> {
```
Comment (siddontang) [inline src/util/config.rs]: s/get_integer_by_string/parse_readable_int/

Comment (siddontang) [inline src/util/config.rs]: please return a parse error here.

**Comment by ngaut** [comment]
PTAL @BusyJay

**Comment by siddontang** [comment]
LGTM

File: src/bin/tikv-server.rs [hunk 8]
```diff
@@ -98,10 +98,20 @@ fn get_integer_value<F>(short: &str,
...
     i.or_else(|| {
-            config.lookup(long).and_then(|v| f(v)).or_else(|| {
-                info!("{}, use default {:?}", long, default);
-                default
-            })
+            config.lookup(long)
+                .and_then(|v| {
+                    if let toml::Value::String(ref s) = *v {
+                        util::config::parse_readable_int(s)
+                            .map_err(|e| info!("{}, {}", long, e))
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Why info? And the message is confusing.

File: src/util/config.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                break;
+            }
+        }
+    }
+
+    let (num, unit) = property.split_at(indx);
+    num.parse::<f64>().map(|f| (f, unit)).or(Err(ParseConfigError))
+}
+
+const UNIT: i64 = 1;
```
Comment (BusyJay) [inline src/util/config.rs]: Can we use `usize` here?

File: src/util/config.rs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+const TIME_MAGNITUDE_2: i64 = 60;
+const MS: i64 = UNIT;
+const SECOND: i64 = MS * TIME_MAGNITUDE_1;
+const MINTUE: i64 = SECOND * TIME_MAGNITUDE_2;
+const HOUR: i64 = MINTUE * TIME_MAGNITUDE_2;
+
+pub fn parse_readable_int(size: &str) -> Result<i64, ParseConfigError> {
+    let (num, unit) = try!(split_property(size));
+
+    match &*unit.to_owned().to_lowercase() {
```
Comment (BusyJay) [inline src/util/config.rs]: Why `to_owned`?

File: src/bin/tikv-server.rs (hunk 8 above)
Comment (overvenus) [inline src/bin/tikv-server.rs]: panic here?

File: src/util/config.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        // file size
+        "kb" => Ok((num * (KB as f64)) as i64),
+        "mb" => Ok((num * (MB as f64)) as i64),
+        "gb" => Ok((num * (GB as f64)) as i64),
+        "tb" => {
+            let mut ret: i64 = 0;
```
Comment (BusyJay) [inline src/util/config.rs]: I think `GB as f64 * num * DATA_MAGNITUDE as f64` is enough.

File: src/bin/tikv-server.rs (hunk 8 above)
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I think panic is acceptable.

Comment (siddontang) [inline src/bin/tikv-server.rs]: +1

File: src/util/config.rs (hunk 11 above)
Comment (siddontang) [inline src/util/config.rs]: f64 may overflow here.

Comment (BusyJay) [inline src/util/config.rs]: How? It may lose precision, but I think it's acceptable.

Comment (overvenus) [inline src/util/config.rs]: Some tests may fail.

Comment (BusyJay) [inline src/util/config.rs]: Which?

Comment (overvenus) [inline src/util/config.rs]: All clear.
After changing type to `uszie`, all overflow warning is gone, and no test fail. Awesome @BusyJay

File: src/util/config.rs [hunk 12]
```diff
@@ -59,51 +59,31 @@ fn split_property(property: &str) -> Result<(f64, &str), ParseConfigError> {
...
-const SECOND: i64 = MS * TIME_MAGNITUDE_1;
-const MINTUE: i64 = SECOND * TIME_MAGNITUDE_2;
-const HOUR: i64 = MINTUE * TIME_MAGNITUDE_2;
+const UNIT: usize = 1;
+const DATA_MAGNITUDE: usize = 1024;
+const KB: usize = UNIT * DATA_MAGNITUDE;
+const MB: usize = KB * DATA_MAGNITUDE;
+const GB: usize = MB * DATA_MAGNITUDE;
+const TB: usize = GB * DATA_MAGNITUDE;
+const PB: usize = TB * DATA_MAGNITUDE;
```
Comment (BusyJay) [inline src/util/config.rs]: To support PB, u64 may be required.

**Comment by BusyJay** [comment]
rest LGTM

File: src/util/config.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+pub fn parse_rocksdb_compression(tp: &str) -> Result<DBCompressionType, ParseConfigError> {
+    match &*tp.to_lowercase() {
+        "no" => Ok(DBCompressionType::DBNo),
+        "snappy" => Ok(DBCompressionType::DBSnappy),
+        "zlib" => Ok(DBCompressionType::DBZlib),
+        "bzip2" => Ok(DBCompressionType::DBBz2),
+        "lz4" => Ok(DBCompressionType::DBLz4),
+        "lz4hc" => Ok(DBCompressionType::DBLz4hc),
+        _ => Err(ParseConfigError),
```
Comment (ngaut) [inline src/util/config.rs]: Error messages should give more details.

File: src/util/config.rs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        "tb" => Ok((num * (TB as f64)) as i64),
+        "pb" => Ok((num * (PB as f64)) as i64),
+
+        // time
+        "ms" => Ok((num * (MS as f64)) as i64),
+        "s" => Ok((num * (SECOND as f64)) as i64),
+        "m" => Ok((num * (MINTUE as f64)) as i64),
+        "h" => Ok((num * (HOUR as f64)) as i64),
+
+        _ => Err(ParseConfigError),
```
Comment (ngaut) [inline src/util/config.rs]: Ditto.

**Comment by siddontang** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Comment by ngaut** [comment]
@overvenus Please rebase master.

**Comment by overvenus** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Pull request merged**